src/HealthChecks.AzureStorage/AzureFileShareHealthCheck.cs

```diff
@@ -1,39 +1,36 @@
-using System.Collections.Concurrent;
 using Azure.Storage.Files.Shares;
 using Microsoft.Extensions.Diagnostics.HealthChecks;
 
 namespace HealthChecks.AzureStorage
 {
     public class AzureFileShareHealthCheck : IHealthCheck
     {
-        private readonly string _connectionString;
-        private readonly string? _shareName;
-
-        private static readonly ConcurrentDictionary<string, ShareClient> _shareClientsHolder = new();
+        private readonly ShareServiceClient _shareServiceClient;
+        private readonly AzureFileShareHealthCheckOptions _options;
 
         public AzureFileShareHealthCheck(string connectionString, string? shareName = default)
+            : this(
+                  ClientCache<ShareServiceClient>.GetOrAdd(connectionString, _ => new ShareServiceClient(connectionString)),
+                  new AzureFileShareHealthCheckOptions { ShareName = shareName })
+        { }
+
+        public AzureFileShareHealthCheck(ShareServiceClient shareServiceClient, AzureFileShareHealthCheckOptions options)
         {
-            _connectionString = connectionString ?? throw new ArgumentNullException(nameof(connectionString));
-            _shareName = shareName;
+            _shareServiceClient = shareServiceClient ?? throw new ArgumentNullException(nameof(shareServiceClient));
+            _options = options ?? throw new ArgumentNullException(nameof(options));
         }
 
         public async Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken = default)
         {
-            try
-            {
-                var shareClient = _shareClientsHolder.GetOrAdd($"{_connectionString}{_shareName}", _ => new ShareClient(_connectionString, _shareName));
+            await _shareServiceClient.GetPropertiesAsync(cancellationToken);
 
-                if (!await shareClient.ExistsAsync(cancellationToken))
```

Comment:
Why did you remove `ExistsAsync`?

Reply:
If the container doesn't exist, `GetProperties` will throw a `RequestFailedException` with a 404 status and subsequently that 404 will be converted to a failed health check. That way, for the typical "healthy" code path, there are only 2 remote calls made to the Azure Storage service instead of 3.

Reply:
For the service being unavailable, the message becomes something like: `"Retry failed after 6 tries. Retry settings can be adjusted in ClientOptions.Retry. (No such host is known. (unknown-blob-service.blob.core.windows.net:443)) ..."`. If the service exists but the container does not, we would get a message like `""The specified container does not exist.\nRequestId:5d9aa1a2-d68b-4b31-8f77-9dad95ac0c60\n<a bunch of request metadata>"`.

At the end of the day, the health check endpoint will still just return the phrase "Unhealthy"

Reply:
ok